iree/compiler/Dialect/HAL/Target/TargetBackend.cpp

```diff
@@ -102,33 +102,85 @@ void TargetBackend::declareTargetOps(IREE::Flow::ExecutableOp sourceOp,
-                            DenseMap<StringRef, Operation *> &targetSymbolMap) {
+// |targetSymbolMap| is updated with the new symbols.
+//
+// If a private symbol in |sourceModuleOp| conflicts with another symbol
+// (public or private) tracked in |targetSymbolMap|, it will be renamed and
+// |privateSymbolOccurrences| will be updated to reflect this.
+//
+// Fails if a public symbol in |sourceModuleOp| conflicts with another public
+// symbol tracked in |targetSymbolMap|.
+static LogicalResult mergeModuleInto(
+    Operation *sourceModuleOp, Operation *targetModuleOp,
+    DenseMap<StringRef, Operation *> &targetSymbolMap,
+    llvm::StringMap<unsigned> &privateSymbolOccurrences) {
   auto &sourceBlock = sourceModuleOp->getRegion(0).front();
   auto &targetBlock = targetModuleOp->getRegion(0).front();
   auto allOps = llvm::to_vector<8>(
       llvm::map_range(sourceBlock, [&](Operation &op) { return &op; }));
+
   for (auto &op : allOps) {
     if (op->hasTrait<OpTrait::IsTerminator>()) continue;
-    if (auto symbolInterface = dyn_cast<SymbolOpInterface>(op)) {
-      if (targetSymbolMap.count(symbolInterface.getName())) {
-        // TODO(scotttodd): compare ops to ensure we aren't copying different
-        // things with the same name.
-        continue;
+    if (auto symbolOp = dyn_cast<SymbolOpInterface>(op)) {
+      auto symbolName = symbolOp.getName();
+
+      // Resolve symbol name conflicts.
+      if (targetSymbolMap.count(symbolName)) {
+        auto targetOp = targetSymbolMap[symbolName];
+
+        if (symbolOp.getVisibility() == SymbolTable::Visibility::Private) {
+          // Private symbols can be safely folded into duplicates or renamed.
+          if (OperationEquivalence::isEquivalentTo(targetOp, op)) {
+            // Optimization: skip over duplicate private symbols.
+            // We could let CSE do this later, but we may as well check here.
+            continue;
+          } else {
+            // Preserve the op but give it a unique name.
+            auto disambiguateName =
+                llvm::formatv("{0}_{1}", symbolName,
+                              privateSymbolOccurrences[symbolName]++)
+                    .str();
+            SymbolTableCollection symbolTable;
+            SymbolUserMap symbolUsers(symbolTable, sourceModuleOp);
+            symbolUsers.replaceAllUsesWith(op, disambiguateName);
+            SymbolTable::setSymbolName(op, disambiguateName);
+          }
+        } else {
+          // The source symbol has 'nested' or 'public' visibility.
+          if (SymbolTable::getSymbolVisibility(targetOp) !=
+              SymbolTable::Visibility::Private) {
+            // Oops! Both symbols are public and we can't safely rename either.
+            // If you hit this with ops that you think are safe to rename, mark
+            // them private.
+            return op->emitError()
+                   << "multiple public symbols with the name: " << symbolName;
```

Comment:
We could also skip linking between executables with conflicting public symbol names. I guess it depends on what sorts of cases where there could be conflicts and if an error would be more appropriate.

Reply:
If this doesnt fail today then reporting error is fine. It is a conflicting symbol, so error seems like a good option.

Reply:
Ok, I expanded on the comment.